### ISSUE TITLE
`NamedTemporaryFile` is incompatible with anything

### ISSUE THREAD

**Issue by karolzlot** [microsoft/pyright#3539]
Not sure if this should be reported in [typeshed](https://github.com/python/typeshed/issues)?

**Describe the bug**
As shown below.

**To Reproduce**
As shown below.

**Expected behavior**
No pyright errors.

**Screenshots or Code**
```python
from typing import BinaryIO
from os import PathLike
from tempfile import NamedTemporaryFile


def render_one(
    output_file_or_path: str | bytes | PathLike[str] | BinaryIO,

):
    print("do something")


tmp_output_file=NamedTemporaryFile()

render_one(tmp_output_file)
```

Result: 
![image](https://user-images.githubusercontent.com/11590960/172035907-39c2d1de-d79c-4cb9-916c-0dea9d2a08b4.png)

**VS Code extension or command-line**
pylance v2022.6.3

**Additional context**
n/d

**Comment by erictraut** [comment]
Pyright is doing the right thing here. The type of the expression `NamedTemporaryFile()` is `_TemporaryFileWrapper[bytes]`. This type isn't compatible with `str`, `bytes`, `PathLike[str]` or `BinaryIO`. It is compatible with `IO[bytes]`, so perhaps you intended to include that in the union of types supported by `render_one`?

`NamedTemporaryFile` and `_TemporaryFileWrapper` are defined [here](https://github.com/python/typeshed/blob/a2ba0c8a00385e62655c7fc714b814eeb2533b5a/stdlib/tempfile.pyi#L38) and [here](https://github.com/python/typeshed/blob/a2ba0c8a00385e62655c7fc714b814eeb2533b5a/stdlib/tempfile.pyi#L184), respectively. If you think these are defined incorrectly, please file a bug in the [typeshed issue tracker](https://github.com/python/typeshed/issues).

I'm going to close this issue because pyright is applying the type information correctly.

**Comment by karolzlot** [comment]
Thanks, what I needed is `str | PathLike[str] | IO[bytes]` instead of `str | bytes | PathLike[str] | BinaryIO`.
